File: 3. Longest Substring Without Repeating Characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: 'str') -> 'int':
        start = 0
        curMax = 0
        used_dict = {}
        for n, char in enumerate(s):
            if (char in used_dict) and (used_dict[char] >= start):  # new char
                start = used_dict[char] + 1  # new start
                used_dict[char] = n  # new to dict
                # if n >= 1 and s[n] == s[n - 1]:
                #     start -= 1
            else:  # in used list
                used_dict[char] = n  # add to dict
                # print('curr max', curMax, 'new max', n - start + 1)
                curMax = max(curMax, n - start + 1)  # new count

            # print('_____________________curr max', curMax, 'new max', n - start + 1)
            # print('start:',s[start], start, '\n')
        return curMax

File: 3. Longest_Substring_Without_Repeating_Characters.cpp
//
// 3. Longest Substring Without Repeating Characters
// Given a string, find the length of the longest substring without repeating characters.
//

#include <iostream>    // std::cout
#include <algorithm>  //std::max
#include <set>
#include <string>

class Solution {
public:
    int lengthOfLongestSubstring(std::string s) {
        int n = s.length();
        int ans = 0;
        for (int i=0; i<n; i++){
            for (int j=i+1; j<=n; j++){
                if (allUnique(s, i, j)){
                    ans = std::max(ans, j-i);
                }
            }
        }
        return ans;
    }

    bool allUnique(std::string s, int start, int end){
        std::set<char>unique_set;
        for (int i=start; i<end; i++){
            auto search = unique_set.find(s[i]);
            if (search != unique_set.end()){
                return false;
            }
            else{
                unique_set.insert(s[i]);
            }
        }
        return true;
    }
};




